yapapi/events.py

```diff
@@ -34,21 +112,128 @@ def exception(self) -> Optional[BaseException]:
 
 @attr.s(auto_attribs=True)
 class JobEvent(Event, abc.ABC):
-    job_id: str
+    job: "Job"
+
+    @property
+    def job_id(self) -> str:
+        return self.job.id
+
+    @property
+    def expires(self) -> datetime:
+        return self.job.expiration_time
+
+    @property
+    def num_offers(self) -> int:
+        return self.job.offers_collected
+
+
+@attr.s(auto_attribs=True)
+class SubscriptionEvent(JobEvent, abc.ABC):
+    subscription: "Subscription"
+
+
+@attr.s(auto_attribs=True)
+class ProposalEvent(JobEvent, abc.ABC):
+    proposal: "OfferProposal"
+
+    @property
+    def prop_id(self) -> str:
+        return self.proposal.id
+
+    @property
+    def provider_id(self) -> str:
+        return self.proposal.issuer
+
+
+@attr.s(auto_attribs=True)
+class AgreementEvent(JobEvent, abc.ABC):
+    agreement: "Agreement"
+
+    @property
+    def agr_id(self) -> str:
+        return self.agreement.id
+
+    @property
+    def provider_id(self) -> str:
+        return self.agreement.cached_details.raw_details.offer.provider_id  # type: ignore
+
+    @property
+    def provider_info(self) -> "NodeInfo":
+        return self.agreement.cached_details.provider_node_info
```

Comment:
Nitpicking, but maybe we should add a comment here about accessing `cached_details` possibly raising a `RuntimeError`?

Reply:
In the current `yapapi` it's just impossible.
In some future version of `yapapi` in will still be impossible unless we make a lot of changes in `agreements_pool`. If we change a lot in `agreements_pool`, there's a big chance we'll also touch this.